MENTAL-MODEL.md

```diff
@@ -0,0 +1,376 @@
+# What is the Payjoin SDK and How does it work?
+
+The Payjoin SDK/`rust-payjoin` is the most well-tested and flexible library for BIP 78 Payjoin and related privacy practices.
+
+The primary crate, `payjoin`, is runtime-agnostic. Data persistence, chain interactions, and networking may be provided by custom implementations or copy the reference `payjoin-client` + bitcoind, `nolooking` + LND integration, or `bitmask-core` + BDK integrations.
+
+The following is a breakdown of the existing documentation and its application to the `payjoin-client` reference implementation.
+
+## Send a PayJoin
+
+The `sender` feature provides the check methods and PSBT data manipulation necessary to send payjoins. Just connect your wallet and an HTTP client. The reference implementation uses `reqwest` and Bitcoin Core RPC. Only a few non-default parameters are required:
+
+```rs
+fn send_payjoin(
+    bitcoind: bitcoincore_rpc::Client,
+    bip21: &str,
+    danger_accept_invalid_certs: bool,
+) -> Result<()>
+```
+
+Default modules including http and a bitcoin wallet may be useful additions to this library.
+
+The `danger_accept_invalid_certs` parameter is used for testing purposes only detailed in sectino 5.
+
+### 1. Parse BIP21 as `payjoin::Uri`
+
+Start by parsing a valid BIP 21 uri having the `pj` parameter. This is the [`bip21`](https://crates.io/crates/bip21) crate under the hood.
+
+```rs
+let link = payjoin::Uri::try_from(bip21)
+    .map_err(|e| anyhow!("Failed to create URI from BIP21: {}", e))?;
+
+let link = link
+    .check_pj_supported()
+    .map_err(|e| anyhow!("The provided URI doesn't support payjoin (BIP78): {}", e))?;
+```
+
+### 2. Construct URI request parameters, a finalized "Original PSBT" paying `.amount` to `.address`
+
+```rs
+let mut outputs = HashMap::with_capacity(1);
+outputs.insert(link.address.to_string(), amount);
+
+let options = bitcoincore_rpc::json::WalletCreateFundedPsbtOptions {
+    lock_unspent: Some(true),
+    fee_rate: Some(Amount::from_sat(2000)), // SHOULD CHANGE TO AN OPTIONAL FEE RATE
+    ..Default::default()
+};
+let psbt = bitcoind
+    .wallet_create_funded_psbt(
+        &[], // inputs
+        &outputs,
+        None, // locktime
+        Some(options),
+        None,
+    )
+    .context("Failed to create PSBT")?
+    .psbt;
+let psbt = bitcoind
+    .wallet_process_psbt(&psbt, None, None, None)
+    .with_context(|| "Failed to process PSBT")?
+    .psbt;
+let psbt = load_psbt_from_base64(psbt.as_bytes()) // SHOULD BE PROVIDED BY CRATE AS HELPER USING rust-bitcoin base64 feature
+    .with_context(|| "Failed to load PSBT from base64")?;
+log::debug!("Original psbt: {:#?}", psbt);
+let pj_params = payjoin::sender::Configuration::with_fee_contribution(
+    payjoin::bitcoin::Amount::from_sat(10000),
+    None,
+);
+```
```

Comment:
I think the current flow works and **it does not need to change to prove the concept**, but it may also be improved with the builder pattern (#19). The builder might take the bip21, an optional amount that differs from that in the bip21, fee rate, a closure to fund and sign an Original PSBT. The result should be a request containing `(original_psbt, optional_parameters)` and state in context.